### PR TITLE
feedback on additional indexes needed

### PR DESCRIPTION
**Pull request by evbogue**
@mixmix's chronological index of posts had me thrilled, so I got excited and created yet another index that allows me to get chronological threads from a message root.

With this I was able to easily create threads within http://github.com/evbogue/mvd

`mvd` also serves a chronological posts feed using @mixmix's latest patch.

What I really want is an index of _everything_ in chronological order, but I couldn't figure out how to do that.

We don't need to merge this, however I want to have a discussion about which indexes would be the best indexes to have available via `ssb-query`.

### CODE REVIEW

**Comment by mixmix** [comment]
For a feed off everything in chronological order we'd need an index like

{ key 'lor', value : [[ 'value',  'timestamp' ]]}

And your query wins have to include something in the published timestamp
slit. A mundane placeholder like { $gt: 0 } would work.


For building threads I think the root index is a neat idea. I have been
using backinks plugin to do this. Like get all the backinks for the root,
then filter fish to the messages which are posts whose root is pointing at
the root msg. Then ssb-sorting.
I think either way we have to ssb-sort because publish time doesn't
necessarily give us casual order over different users (e.g. I am offline
for a week but have posted a root to the root msg 6 days into my solitude.
Casually I think it's right after the initial root msg, so when I rejoin it
should be sorted there, even though my post is temporally late... Taking
about this is tricky!)

**Comment by evbogue** [comment]
@mixmix Yah, this is tricky. I can see benefits to both the causal ordered approach _and_ the timestamp ordered approach. But they are clearly different, so it's probably best to stick with causual ordering for threads, so that they appear similar in different clients.

**Comment by evbogue** [comment]
@mixmix I got it working! https://github.com/evbogue/ssb-query/commit/73d36e800e42129922836080ff09d3e2d0cc8a65 

This is awesome. The only downside is with these two new indexes, my query folder is now up to 78 megabytes.

**Comment by mixmix** [comment]
yeah it starts to add up. There are ~~2~~ 3 approaches we could take here : 
- fuck it, 78MB is tiny
- pause and think carefully about the use cases we are going to need these for, keeping in mind _partial matches_ on some indexes still get decent perf boosts over full read of db
- split `ssb-query` into several indexes so people can pick and mix what they need
  - this one seems ugly, but it would also make it really easy to extend and upgrade different plugins ... ? e.g. perhaps channel-related queries should really be all collected into a plugin that is really all about channels (which is a module I feel is needed), and a plugin about type etc. The disadvantage is would wouldn't get smart index selection ...

**Comment by mixmix** [comment]
@evbogue  (do you need to mention to trigger good notificaitons I wonder)

**Comment by dominictarr** [comment]
@mixmix my plan is to refactor ssb-query (i mean, flumeview-query) so that it is decoupled from the actual indexes. So that indexes provided by other plugins can be used by ssb-query, not just indexes that ssb-query owns. This achives two goals: plugins can specify indexes they need, and I do not have veto power over indexes used by ssb-query.

**Comment by evbogue** [comment]
@dominictarr Good plan.

**Comment by mixmix** [comment]
Sounds great

So in the meantime, I think we should pile in the indexes. I'm all for
merging this PR!

**Comment by mmckegg** [comment]
I'm just coming up against this exact issue in PATCHTRON. I want to be able to avoid creating my own indexes as much as possible, but I need to be able to get the feed in "asserted" order.

The only change I'd recommend to this would be to [do what backlinks does and index `rts`](https://github.com/ssbc/ssb-backlinks/blob/master/index.js#L8) instead of `value.timestamp`. 

`rts` [in this context](https://github.com/ssbc/ssb-backlinks/blob/master/emit-links.js#L28) is `Math.min(msg.timestamp, msg.value.timestamp)`. The advantage to this is that people with incorrectly set system clocks can't accidentally (or not) pin their posts to the top of someone else's feed. 

Basically if this is not merged, I'll be running a forked version of ssb-query in PATCHTRON, and the sadness of incompatible/duplicate indexes will continue 😢

**Comment by evbogue** [comment]
@mmckegg Yes, I'm having this issue in http://github.com/evbogue/mvd too.

I ended up forking `ssb-query` in `mvd`. I'd prefer to be able to depend on it. 

I'm thinking about how to solve this...

I can probably depend on your `rts` index instead of indexing by timestamp.

I had an issue where someone was pinning their posts to the top of the feed because they'd posted 5 months in the future. I ended up solving this by discarding posts from the future, but your solution sounds smarter than mine.
